README.md
# ScrAPI

A simple API to scrape specific HTML data.

---

## Install

`npm i`

### HTTPS

Some experiences will complain if the App isn't run over https. To get that
wired up, follow the below instructions.

- Run `./bin/gen-certs.sh "App Name"`
   - This'll create a `certs.app-name` folder with files like
     ```sh
     /certs.app-name
       app-name.crt
       app-name.key
       app-name-CA.crt
       app-name-CA.key
     ```
- The non-`-CA` files will be used for the App. When starting the App via Node
or Docker, you'll need to set this environment variable
   ```sh
   `NODE_EXTRA_CA_CERTS="$PWD/certs/app-name.crt"`
   ```
   - Note that `$PWD` expands to an absolute file path.
   - The App automatically determines the `.key` file so long as the `.key` & `.crt`
   files have the same name.
- Install the Certificate Authority in **Chrome**:
   - Settings > In the top input, filter by `cert` > Click `Security`
   - Click on `Manage certificates`
   - Go the `Trusted Root Certification Authorities` tab
   - Choose `Import`
   - Find the `certs/app-name-CA.crt` file, and add it.
   - If the cert doesn't seem to be working, try in Incognito first. If it's
   working there, then just restart Chrome to get it to work in non-Incognito.
- Install the Certificate Authority in **Firefox**:
   - Options > In the top input, filter by `cert` > Click `View Certificates...`
   - Go to the `Authorities` tab
   - Click on `Import`
   - Find the `certs/app-name-CA.crt` file, and add it.
   - Check `Trust this CA to identify websites`.
- Install the Certificate Authority on **Android**:
   - Copy the CA `.crt` & `.key` to the device
   - Go to `Settings` > `Security` > Click on `Install from storage`
   - Select the `.crt` file
   - Give it a name

---

## Start

```sh
# Standard Server
npm run start

# Development Server
npm run start:dev

# Via Docker
docker-compose up --abort-on-container-exit
```

For more info on how the `config` for the VPN is set up, visit: https://github.com/the0neWhoKnocks/docker-vpn

---

## Request Data

**Accepted Params**
| Param | Optional | Description |
| ----- | -------- | ----------- |
| `headers` | `true` | Any extra headers you may need to send along to the `url` |
| `selectors` | `false` | A pipe `|` delimited list of CSS selectors. |
| `ua` | `true` | Whether or not to pass the request's User-Agent along. |
| `url` | `false` | The URL to the page you want to scrape. |

```js
const url = new URL('https://localhost:3000/api');
url.search = new URLSearchParams({
  // Any extra headers that may be required for the page
	headers: encodeURIComponent(JSON.stringify({
    Host: 'site.com',
    Referer: 'https://site.com/page/',
	})),
  // A list of CSS selectors separated by a pipe `|` character
	selectors: encodeURIComponent('meta[itemprop="image"]|img[alt="Google"]|form[action="/search"]|.bad-selector'),
	ua: true,
  // The URL of the page you want to scrape data from
	url: 'https://google.com/',
}).toString();

fetch(url)
  .then(resp => resp.json())
  .then((resp) => {
    console.log(resp);
  })
  .catch((err) => {
    console.error(err);
  });
```

---

## Releasing

1. Prep the repo
   - Ensure the info in `./bin/releaseConfig.sh` is correct for your repo.
   - Create a `CHANGELOG.md` if one does not already exist. The contents should
   be this.
      ```md
      # Changelog
      ---

      ```
1. Run `./bin/release.sh` and follow the steps.
   - To skip having to enter a Docker password every time, create a
   `.dockercreds` file in the same directory as `release.sh` and put your
   password in there.
   - In order to automatically convert git tags to actual releases you'll need
   to generate a [Personal Access Token](https://github.com/settings/tokens),
   and add it to your global git config `git config --global github.token <YOUR_TOKEN>`.

If something happens during the final stage of the release, you'll have to
manually reset some things.
```sh
# Reset the last commit
git reset --soft HEAD~1
# Verify that just release files will be reset. You should just see:
# - `CHANGELOG.md`
# - `package-lock.json`
# - `package.json`
git status
# If the above is good, unstage those changes
git reset
# Reset changed files
git checkout -- CHANGELOG.md package.json package-lock.json
# Check if a git tag was created
git tag | cat
# If so, remove it
git tag -d <TAG_NAME>
```


CHANGELOG.md
# Changelog
---

## v1.6.0

**Dev-Ops**
- [b2368dc](https://github.com/the0neWhoKnocks/scrapi/commit/b2368dc):  Updated release script

**Features**
- [bb1c941](https://github.com/the0neWhoKnocks/scrapi/commit/bb1c941):  Ability to add custom headers

---

## v1.5.0

**Bugfixes**
- [3adc59c](https://github.com/the0neWhoKnocks/scrapi/commit/3adc59c):  Multiple results and errors

**Dev-Ops**
- [238da5d](https://github.com/the0neWhoKnocks/scrapi/commit/238da5d):  Container wouldn't build
- [c8e47eb](https://github.com/the0neWhoKnocks/scrapi/commit/c8e47eb):  Brought in new release script

---

## v1.4.0

**Features**
- [222fa4b](https://github.com/the0neWhoKnocks/scrapi/commit/222fa4b) - Added VPN support
- [8b44208](https://github.com/the0neWhoKnocks/scrapi/commit/8b44208) - User-Agent
- [d6fb58e](https://github.com/the0neWhoKnocks/scrapi/commit/d6fb58e) - Added inspect middleware

**Uncategorized**
- [d6870b9](https://github.com/the0neWhoKnocks/scrapi/commit/d6870b9) ops: Release not created in GH when double quotes in commit message

---

## v1.3.0

**Features**
- [7beef50](https://github.com/the0neWhoKnocks/scrapi/commit/7beef50) - Treat failure like success

---

## v1.2.0

**Features**
- [25222e8](https://github.com/the0neWhoKnocks/scrapi/commit/25222e8) - Added CORS headers

**Misc. Tasks**
- [3d3eb84](https://github.com/the0neWhoKnocks/scrapi/commit/3d3eb84) - Update release URL to HTTPS

---

## v1.1.0

**Features**
- [6f4288f](https://github.com/the0neWhoKnocks/scrapi/commit/6f4288f) - Added HTTPS support

---


watcher.js
const nodemon = require('nodemon');
const { PORT } = require('./src/constants');

const port = +process.env.PORT || PORT;
const LOG_PREFIX = '[WATCHER]';
let httpModule;
let protocol = 'http';

if (process.env.NODE_EXTRA_CA_CERTS) {
  httpModule = require('https');
  protocol = 'https';
}
else {
  httpModule = require('http');
}

const checkServer = () => new Promise((rootResolve, rootReject) => {
  let count = 0;
  const check = () => new Promise((resolve, reject) => {
    setTimeout(() => {
      const serverAddress = `${protocol}://localhost:${port}`;
      const opts = {};
      
      if (protocol === 'https') {
        // NOTE - Depending on your Dev env, your self-signed certs may
        // throw this error `UNABLE_TO_VERIFY_LEAF_SIGNATURE` during Server
        // restart. Not sure why it doesn't happen on start of the Server, but
        // this will get around that issue (which is fine in development, not Prod).
        opts.rejectUnauthorized = false;
      }

      console.log(`${LOG_PREFIX} Pinging ${serverAddress}`);
      httpModule
        .get(serverAddress, opts, (res) => resolve(res))
        .on('error', (err) => reject(err));
    }, 1000);
  });
  const handleError = (err) => {
    if (count < 3) {
      ping();
      count++;
    }
    else {
      console.log(err)
      rootReject();
    }
  };
  const handleSuccess = () => { rootResolve(); };
  const ping = () => {
    check()
      .then(handleSuccess)
      .catch(handleError);
  };

  ping();
});

nodemon({
  delay: 500,
  script: './src/server',
  watch: [
    './src/server/**/*.js'
  ],
})
  .on('restart', () => {
    console.log(`${LOG_PREFIX} Server restarting because file(s) changed`);

    checkServer()
      .then(() => {
        console.log('Server has fully started');
      })
      .catch(() => {
        console.log("Couldn't detect the server, a manual reload may be required");
      });
  });

function killWatcher(evType) {
  console.log(`${LOG_PREFIX} Killing watcher (${evType})`);
  nodemon.emit('quit');
  process.exit(0);
}

process.on('SIGINT', killWatcher.bind(null, 'SIGINT'));
process.on('SIGTERM', killWatcher.bind(null, 'SIGTERM'));
process.on('SIGUSR2', killWatcher.bind(null, 'SIGUSR2'));

src/server/middleware/api.js
const axios = require('axios');
const cheerio = require('cheerio');

const apiMiddleware = (req, resp, next) => {
  if (req.url.startsWith('/api')) {
    const { headers, selectors, ua, url } = req.query;
    const userAgent = ua && req.headers && req.headers['user-agent'];
    const additionalHeaders = headers && JSON.parse(decodeURIComponent(headers)) || undefined;
    
    console.log(`[LOAD] "${url}"`);
    
    resp.setHeader('Access-Control-Allow-Origin', '*');
    resp.setHeader('Access-Control-Allow-Headers', 'Origin, X-Requested-With, Content-Type, Accept');
    
    const parseResponse = (payload) => {
      if (
        (!payload && !payload.response)
        || payload instanceof Error
      ) {
        const status = payload.status || payload.response.status;
        resp.error(payload.stack, status);
      }
      else {
        if (payload.status === undefined) payload = payload.response;
        
        const results = [];
        const html = payload.data;
        const $ = cheerio.load(html);
        const _selectors = decodeURIComponent(selectors).split('|');
        
        // console.log($.root().html());

        _selectors.forEach(selector => {
          console.log(`  [FIND] Selector: "${selector}"`);

          const nodeData = $(selector);

          if (nodeData && nodeData.length) {
            nodeData.each((ndx, node) => {
              results.push({
                attr: node.attribs,
                html: $(node).html(),
                text: $(node).text(),
              });
            });
          }
        });
        
        resp.json({ results });
      }
    };
    
    const opts = {};
    
    if (userAgent || additionalHeaders) opts.headers = {};
    
    if (userAgent) opts.headers['User-Agent'] = userAgent;
    
    if (additionalHeaders) {
      Object.keys(additionalHeaders).forEach((header) => {
        opts.headers[header] = additionalHeaders[header];
      });
    }
    
    if (Object.keys(opts).length) {
      const formattedOpts = JSON.stringify(opts, null, 2)
        .split('\n')
        .map(line => `  ${line}`)
        .join('\n');
      console.log(`  [REQ_OPTS]\n${formattedOpts}`);
    }
    
    axios.get(url, opts).then(parseResponse).catch(parseResponse);
  }
  else next();
};

module.exports = apiMiddleware;

src/server/index.js
const compression = require('compression');
const polka = require('polka');
const { PORT } = require('../constants');
const apiMiddleware = require('./middleware/api');
const heartbeatMiddleware = require('./middleware/heartbeat');
const inspectMiddleware = require('./middleware/inspect');

const port = process.env.PORT || PORT;
const middleware = [
  compression({ threshold: 0 }),
  inspectMiddleware(),
  (req, res, next) => {
    res.error = function error(msg, status = 500) {
      res.status(status).json({ error: msg });
    }
    res.json = function json(data = {}) {
      res.setHeader('Content-Type', 'application/json');
      res.end(JSON.stringify(data));
    }
    res.status = function status(code = 200) {
      res.statusCode = code;
      return res;
    }
    next();
  },
  heartbeatMiddleware,
  apiMiddleware,
];
const { handler: requestHandler } = polka().use(...middleware);

let httpModule;
let protocol = 'http';
let server;

if (process.env.NODE_EXTRA_CA_CERTS) {
  const { readFileSync } = require('fs');
  const cert = readFileSync(process.env.NODE_EXTRA_CA_CERTS, 'utf8');
  const key = readFileSync(process.env.NODE_EXTRA_CA_CERTS.replace('.crt', '.key'), 'utf8');
  
  httpModule = require('https');
  protocol = 'https';
  server = httpModule.createServer({ cert, key }, requestHandler);
}
else {
  httpModule = require('http');
  server = httpModule.createServer(requestHandler);
}

function startServer() {
  server.listen(PORT, err => {
    if (err) console.log('error', err);
    else {
      console.log(`Server running at: ${protocol}://localhost:${port}`);
      
      if (process.env.VPN_ENABLED) {
        const axios = require('axios');
        axios.get('https://api.ipify.org/', { timeout: 3000 })
          .then(resp => {
            console.log(`Public IP: "${resp.data}"`);
          })
          .catch(err => {
            console.log(err);
            process.exit(1);
          });
      }
    }
  });
}

if (process.env.VPN_ENABLED) {
  const { networkInterfaces } = require('os');
  const maxSeconds = 60;
  let count = 1;
  
  console.log('[WAIT] For VPN to connect');
  
  const waitForConnection = setInterval(() => {
    const nets = networkInterfaces();
    
    for (const name of Object.keys(nets)) {
      for (const net of nets[name]) {
        if (net.family === 'IPv4' && !net.internal) {
          if (name.startsWith('tun')) {
            clearInterval(waitForConnection);
            startServer();
          }
        }
      }
    }
    
    if (count === maxSeconds) {
      clearInterval(waitForConnection);
      console.log(`[ERROR] It's been ${maxSeconds} seconds, and the VPN hasn't connected. Giving up.`);
      process.exit(1);
    }
    count += 1;
  }, 1000);
}
else startServer();
